keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,728 @@
+
+Maximum number of keys is estimated around 4-8, but it is not set as a hard limit, because setting a hard limit will make the behavior inconsistent (which key can't be added depends on the order of addition).
+
+### Other Design Considerations
+
+#### Behavior with `ownerReference`
+
+Lien itself only blocks a deletion request to an object that is added to.
+However, it may not be clear how lien behaves when used with [`ownerReference`](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents).
+Therefore, this section describes the behavior.
+
+In summary, lien will block cascading deletion, but would not block deletion of any dependent resources individually.
+
+For example, let's assume that we have Deployment A which manages Replicaset B which manages Pod C.
+In this situation, ownerRefernces are set from Pod C to Replicaset B and from Replicaset B to Deployment A.
+
+If a lien is set to Deployment A, only a deletion request to Deployment A is blocked.
+Users can still request to delete Replicaset B and Pod C directly, but they can't request to delete them through Deployment A.
+
+If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
+Users can still request to delete Deployment A and Pod C directly.
+When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.
+The behaviors are as follows:
+
+- Foreground cascading deletion:
+  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
+  - otherwise: Deployment A is deleted immediately
+- Background cascading deletion: Deployment A is deleted immediately
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked until all other resources without liens are deleted.
+This means a resource with lien will eventually be deleted.
+Users can still request to delete its namespace or other objects in the namespace.
+
+When deletion of the namespace is requested, the resource deletion in the deleted namespace is done by namespace lifecycle controller in below two steps:
+1. Delete all resources (as a result, only resources without lien are deleted)
+2. Force delete all resources (as a result, all resources are deleted regardless of lien)
```

Comment:
So putting a lien on a pod and then deleting the namespace will ultimately override the lien?  Is that true with finalizers, too?  I think the behavior should be the same

Reply:
It's not true for finalizers - finalizers has to be explicitly deleted.

I personally agree with the sentiment to unify their behavior, but IIRC it was @deads2k concern that adding a new mechanism that could break your automation (if e.g. namespace deletion doesn't delete the objects even though it previous was because now someone added a lien). I don't think it fully persuades me, but it's definitely a valid one.

Reply:
Is this the main unresolved issue here?

Reply:
We could write up the concerns and agree not to move to beta until there's evidence that these are covered.

I'm thinking that there will be 2 cases:
1. same-namespace liens
1. liens between a namespaced object and something not in that namespace

the second case seems like a good example of something that might block namespace deletion indefinitely; for the first case, you'd hope to see the controllers able to make progress and perhaps even to reflect that progress back via Events / status / whatever.